src/Emulator/Peripherals/Peripherals/GPIOPort/LPC43xx_GPIO.cs

```diff
@@ -0,0 +1,181 @@
+        {
+            var portNumber = (offset & 0x1F) >> 2;
+
+            UInt32 newValue = pin[portNumber];
+
+            switch((Registers)(offset & 0xFFE0))
+            {
+                case Registers.GPIO_DIR:
+                    dir[portNumber] = value;
+                    break;
+
+                case Registers.GPIO_MASK:
+                    mask[portNumber] = value;
+                    break;
+
+                case Registers.GPIO_PIN:
+                    UpdateOutputs((byte) portNumber, value);
+                    break;
+
+                case Registers.GPIO_SET:
+                    BitHelper.OrWith(ref newValue, value, 0, 32);
+                    UpdateOutputs((byte) portNumber, newValue);
+                    break;
+
+                case Registers.GPIO_CLR:
+                    BitHelper.AndWithNot(ref newValue, value, 0, 32);
+                    UpdateOutputs((byte) portNumber, newValue);
+                    break;
+
+                case Registers.GPIO_NOT:
+                    BitHelper.XorWith(ref newValue, value, 0, 32);
+                    UpdateOutputs((byte) portNumber, newValue);
+                    break;
+
+                default:
+                    this.LogUnhandledWrite(offset, value);
+                    break;
+            }
+            return;
+        }
+
+        public override void OnGPIO(int number, bool value)
+        {
+            base.OnGPIO(number, value);
+            Connections[number].Set(value);
+        }
+
+        public override void Reset()
+        {
+            base.Reset();
+            for(var index = 0; index < NumberOfPorts; index++)
+            {
+                this.dir[index] = 0;
+                this.pin[index] = 0;
+                this.mask[index] = 0;
+            }
+        }
+
+        private UInt32 ReadInputs(byte port)
+        {
+            var returnValue = 0u;
+            for(var bit = 0; bit < 32; bit++)
+            {
+                BitHelper.SetBit(ref returnValue,(byte) bit, State[32 * port + bit]);
+            }
+            return returnValue;
+        }
+
+        private void UpdateOutputs(byte port, UInt32 value)
+        {
+            UInt32 changes = pin[port] ^ value;
+            pin[port] = value;
+
+            for(byte bit = 0; bit < 32; bit++)
+            {
+                if(BitHelper.IsBitSet(dir[port], bit) && (BitHelper.IsBitSet(changes, bit)))
+                {
+                    if(BitHelper.IsBitSet(pin[port], bit))
+                    {
+                        Connections[32 * port + bit].Set();
+                        State[32 * port + bit] = true;
+                    }
+                    else
+                    {
+                        Connections[32 * port + bit].Unset();
+                        State[32 * port + bit] = false;
+                    }
+                }
+            }
+        }
+
+        private UInt32[] pin;
+        private UInt32[] dir;
```

Comment:
An example of what I mean (written from the top of my head, so might not be 100% accurate):
```
private class Pin
{
    [...]
    public bool State { get; set; } // describes the current state
    public bool MaskedState => State && Mask;
    public bool Mask { get; set; }
    public Direction Direction { get; set; }
}

private class Port
{
    public Port(int numberOfPins)
    {
        pins = new Pin[numberOfPins];
        // initialization of pins
    }

    [...]

    private readonly Pin[] pins;
}
```

That way you just need a single array of `Port`s.
Additionally you can move all port-related logic, e.g., reading a value to a method in `Port` class.
What's more you can link (e.g., by passing an id via constructor's parameter) `Pin` with `Connections` and `state`, so that changing any of the pin properties (state, mask, direction) automatically updates the proper output.